paasta_tools/docker_wrapper.py

```diff
@@ -84,31 +89,141 @@ def generate_hostname(fqdn, mesos_task_id):
 
 
+def get_cpumap():
+    # Return a dict containing the core numbers per physical CPU
+    core = 0
+    cpumap = {}
+    try:
+        with open('/proc/cpuinfo', 'r') as f:
+            for line in f:
+                m = re.match('physical\sid.*(\d)', line)
+                if m:
+                    cpuid = int(m.group(1))
+                    if cpuid not in cpumap:
+                        cpumap[cpuid] = []
+                    cpumap[cpuid].append(core)
+                    core += 1
+    except IOError:
+        logging.warning('Error while trying to read cpuinfo')
+        pass
+    return cpumap
+
+
+def get_numa_memsize(nb_nodes):
+    # Return memory size in mB per NUMA node assuming memory is split evenly
+    # TODO: calculate and return real memory map
+    try:
+        with open('/proc/meminfo', 'r') as f:
+            for line in f:
+                m = re.match('MemTotal:\s*(\d)kb', line)
+                if m:
+                    return int(m) / 1024 / int(nb_nodes)
+    except IOError:
+        logging.warning('Error while trying to read meminfo')
+        pass
+    return 0
+
+
+def arg_collision(new_args, current_args):
+    # Returns True if one of the new arguments is already in the
+    # current argument list.
+    cur_arg_keys = []
+    for c in current_args:
+        cur_arg_keys.append(c.split('=')[0])
+    return bool(set(new_args).intersection(set(cur_arg_keys)))
+
+
+def extract_float(f):
+    # Ensure we return a float. If input is invalid we return 0.0
+    try:
+        return float(f)
+    except (ValueError, TypeError):
+        logging.warning('Could not read %s as a float' % str(f))
+        return 0.0
+
+
+def is_numa_enabled():
+    if os.path.exists('/proc/1/numa_maps'):
+        return True
+    else:
+        logging.warning('The system does not support NUMA')
+        return False
+
+
+def append_cpuset_args(argv, env_args):
+    # Enable log messages to stderr
+    logging.basicConfig(stream=sys.stderr, level=logging.INFO)
+
+    try:
+        pinned_numa_node = int(env_args.get('PIN_TO_NUMA_NODE'))
+    except (ValueError, TypeError):
+        logging.error('Could not read PIN_TO_NUMA_NODE value as an int: {}'.format(
+            env_args.get('PIN_TO_NUMA_NODE')))
+        return argv
+
+    cpumap = get_cpumap()
+
+    if len(cpumap) < 1:
+        logging.error('Less than 2 physical CPU detected')
+        return argv
+
+    if pinned_numa_node not in cpumap:
+        logging.error('Specified NUMA node: {} does not exist on this system'.format(
+            pinned_numa_node))
+        return argv
+
+    if arg_collision(['--cpuset-cpus', '--cpuset-mems'], argv):
+        logging.error('--cpuset options are already set. Not overriding')
+        return argv
+
+    if not is_numa_enabled():
+        logging.error('Could not detect NUMA on the system')
+        return argv
+
+    if len(cpumap[pinned_numa_node]) < extract_float(env_args.get(
+            'MARATHON_APP_RESOURCE_CPUS')):
```

Comment:
Can you extract this :196 and :201 into a extract_mem_requirement and extract_cpu_requirement and make that do the logic of extracting the right variable and float'ing it? My reasoning is that we need more than marathon support (Eventually) and having a function that encapsulates *how* we extract these values will make it tidy and testable.

In other words, I don't want this function `append_cpuset_args` to "know" about marathon at all.